staff/acct/sorry/sorry

```diff
@@ -1,156 +1,183 @@
+    # Own sorry file
+    if (os.path.isfile(file_path)):
+        return file_path
+
+    # Default sorry files
+    if (os.path.isfile(reason_path + file_path)):
+        return reason_path + file_path
+
+    raise FileNotFoundError("FATAL: Sorry file \"{}\" does not exist. Exiting.".format(file_path))
+
+
+def sorry_user(username, sorry_file):
+    attributes = ocflib.account.search.user_attrs(username)
+    if attributes is None:
+        raise ValueError('User not found.')
+
+    rootstaffer = real_user()
+    kerberos_credentials(rootstaffer)
+
+    user_dir = attributes['homeDirectory']
+
+    # Workaround for ldap3 1.x.x machines
+    if isinstance(user_dir, list):
+        user_dir = user_dir[0]
+
+    print('Copying sorry file and making .oldshell file')
+
+    if not (os.path.isfile(user_dir + '/.oldshell')):
+        oldshell = attributes['loginShell']
+        # Workaround for ldap3 1.x.x machines
+        if isinstance(oldshell, list):
+            oldshell = user_dir[0]
+
+        oldshell_file = open(user_dir + '/.oldshell', 'w')
+        oldshell_file.write(oldshell)
+        oldshell_file.close()
+
+    shutil.copy2(sorry_file, user_dir + '/.sorry')
+    os.chmod(user_dir + '/.sorry', 0o400)
+    shutil.chown(user_dir + '/.sorry', user=username, group='ocf')
+
+    print("Changing user's shell to sorry shell")
+
+    ocflib.infra.ldap.modify_ldap_entry(ocflib.account.utils.dn_for_username(username),
+                                        {'loginShell': sorry_shell, 'gidNumber': grp.getgrnam('sorry').gr_gid})
+
+    http_dir = '/services/http/users/{0}/{1}'.format(username[0], username)
+    print('Changing permissions on httpdir to 000')
+    if (os.path.isdir(http_dir)):
+        os.chmod(http_dir, 0o000)
+
+    print('Changing permissions on home directory to 500')
+    os.chmod(user_dir, 0o500)
+
+    # WTF is final system check for?
+    # log(rootstaffer, username)
+    send_log_mail(rootstaffer, username, open(reason_path, 'r').read())
+
+    user_mail = get_user_real_mail(username)
+    if user_mail:
+        send_sorry_mail(user_mail, username, open(reason_path, 'r').read())
+    kerberos_destroy()
+
+# Returns the real user who is running this script
+
+
+def real_user():
+    if 'SUDO_USER' in os.environ and os.environ['SUDO_USER'] != 'root':
+        return os.environ['SUDO_USER']
+    else:
+        print("The sorry.log is much more useful when it logs who you are rather than simply 'root'.")
+        username = input('Please enter your username: ')
+        return username
+
+
+def kerberos_credentials(staff_name):
+    if ('SORRY_KRB5CCNAME' in os.environ):
+        print('SORRY_KRB5CCNAME set in environment. Assuming this file contains current admin credentials.')
+        os.environ['KRB5CCNAME'] = os.environ['SORRY_KRB5CCNAME']
+    else:
+        print('Your are {0}.'.format(staff_name))
+        if (subprocess.call([shutil.which('kinit'), '{0}/admin'.format(staff_name)]) != 0):
```

Comment:
Ideally, if we're already authenticated with an admin principal, we shouldn't call kinit. `klist --json` will give you a JSON dump of the current credential cache. If it contains the k/v pair `"principal": "username/admin@OCF.BERKELEY.EDU"`, then we don't need to do kinit. This could look something like this:

```python
import json
...
klist = subprocess.run(['sudo', '-u', staff_name, 'klist', '--json'], stdout=subprocess.PIPE)
if klist.returncode == 0:
    cache_info = json.loads(klist.stdout.decode())
    if cache_info.get('principal') == '{0}/admin@OCF.BERKELEY.EDU'.format(staff_name):
        print('Reading from credential cache.')
        return
... #continue to kinit
```

The `sudo` bit is necessary because this script is running as root but `klist` needs to run as the user.